### PR TITLE
Upgrade prettier

### PR DESCRIPTION
**Pull request by BarryThePenguin**
✨ `prettier@2` ✨ 

Hiding white-space may help 👀 https://github.com/emberjs/ember.js/pull/19154/files?diff=split&w=1

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -5,13 +5,12 @@ module.exports = {
   parser: 'babel-eslint',
   extends: [
     'eslint:recommended',
-    'prettier',
     'plugin:import/errors',
     'plugin:qunit/recommended',
+    'plugin:prettier/recommended',
```
Comment (BarryThePenguin) [inline .eslintrc.js:10]: This is the recommended configuration https://github.com/prettier/eslint-plugin-prettier#recommended-configuration

It does three things..

- Enables `eslint-plugin-prettier`.
- Sets the `prettier/prettier` rule to `"error"`.
- Extends the `eslint-config-prettier` configuration.

File: .eslintrc.js [hunk 2]
```diff
@@ -55,6 +53,8 @@ module.exports = {
       files: [ '**/*.ts' ],
 
       parser: '@typescript-eslint/parser',
+      plugins: ['@typescript-eslint'],
+      extends: ['plugin:@typescript-eslint/eslint-recommended', 'prettier/@typescript-eslint'],
```
Comment (BarryThePenguin) [inline .eslintrc.js:57]: [`plugin:@typescript-eslint/eslint-recommended`](https://github.com/typescript-eslint/typescript-eslint/blob/v2.33.0/packages/eslint-plugin/src/configs/eslint-recommended.ts) turns off all the `eslint` rules that conflict with `@typescript-eslint`

`prettier/@typescript-eslint` turns off the `@typescript-eslint` rules that conflict with prettier

File: .eslintrc.js [hunk 3]
```diff
@@ -55,6 +53,12 @@ module.exports = {
       files: [ '**/*.ts' ],
 
       parser: '@typescript-eslint/parser',
+      plugins: ['@typescript-eslint'],
+      extends: [
+        'plugin:@typescript-eslint/eslint-recommended',
+        'plugin:import/typescript',
+        'prettier/@typescript-eslint'
+      ],
```
Comment (BarryThePenguin) [inline .eslintrc.js:61]: [`plugin:@typescript-eslint/eslint-recommended`](https://github.com/typescript-eslint/typescript-eslint/blob/v2.33.0/packages/eslint-plugin/src/configs/eslint-recommended.ts) turns off all the `eslint` rules that conflict with `@typescript-eslint`

[`plugin:import/typescript`](https://github.com/benmosher/eslint-plugin-import#typescript) [includes configuration](https://github.com/benmosher/eslint-plugin-import/blob/v2.20.2/config/typescript.js) when using `eslint-plugin-import` with TypeScript

`prettier/@typescript-eslint` turns off the `@typescript-eslint` rules that conflict with `prettier`

File: .eslintrc.js [hunk 4]
```diff
@@ -55,6 +53,12 @@ module.exports = {
       files: [ '**/*.ts' ],
 
       parser: '@typescript-eslint/parser',
+      plugins: ['@typescript-eslint'],
+      extends: [
+        'plugin:@typescript-eslint/eslint-recommended',
```
Comment (BarryThePenguin) [inline .eslintrc.js:58]: [`plugin:@typescript-eslint/eslint-recommended`](https://github.com/typescript-eslint/typescript-eslint/blob/v2.33.0/packages/eslint-plugin/src/configs/eslint-recommended.ts) turns off all the `eslint` rules that conflict with `@typescript-eslint`

File: .eslintrc.js [hunk 5]
```diff
@@ -55,6 +53,12 @@ module.exports = {
...
+        'plugin:import/typescript',
```
Comment (BarryThePenguin) [inline .eslintrc.js:59]: [`plugin:import/typescript`](https://github.com/benmosher/eslint-plugin-import#typescript) [includes configuration](https://github.com/benmosher/eslint-plugin-import/blob/v2.20.2/config/typescript.js) when using `eslint-plugin-import` with TypeScript

File: .eslintrc.js [hunk 6]
```diff
@@ -55,6 +53,12 @@ module.exports = {
...
+        'prettier/@typescript-eslint'
```
Comment (BarryThePenguin) [inline .eslintrc.js:60]: `prettier/@typescript-eslint` turns off the `@typescript-eslint` rules that conflict with `prettier`

File: .eslintrc.js [hunk 7]
```diff
@@ -65,11 +69,11 @@ module.exports = {
       rules: {
         // the TypeScript compiler already takes care of this and
         // leaving it enabled results in false positives for interface imports
-        'no-dupe-class-members': 'off',
         'no-unused-vars': 'off',
-        'no-undef': 'off',
```
Comment (BarryThePenguin) [inline .eslintrc.js:70]: These two rules are already disabled with https://github.com/typescript-eslint/typescript-eslint/blob/v2.20.0/packages/eslint-plugin/src/configs/eslint-recommended.ts

File: bin/build-for-publishing.js [hunk 8]
```diff
@@ -87,7 +87,7 @@ Promise.resolve()
   .then(
     // eslint-disable-next-line
     () => console.log('build-for-publishing completed successfully!'),
-    error => {
+    (error) => {
```
Comment (chancancode) [inline bin/build-for-publishing.js:90]: Can we change `arrowParens` back to `"avoid"`, if only just minimize the diff?

Comment (BarryThePenguin) [reply bin/build-for-publishing.js:90]: Ah! Sure, I didn't realise it was still an option. I thought they removed it in v2

Comment (BarryThePenguin) [reply bin/build-for-publishing.js]: Ok, done 👍 I was thinking of [Always add a space after the `function` keyword](https://prettier.io/blog/2020/03/21/2.0.0.html#always-add-a-space-after-the-function-keyword-3903httpsgithubcomprettierprettierpull3903-by-j-f1httpsgithubcomj-f1-josephfrazierhttpsgithubcomjosephfrazier-sosukesuzukihttpsgithubcomsosukesuzuki-thorn0httpsgithubcomthorn0-7516httpsgithubcomprettierprettierpull7516-by-bakkothttpsgithubcombakkot)

File: .eslintrc.js (hunk 4 above)
Comment (chancancode) [reply .eslintrc.js:58]: Do these arrays get merged with the outer config, or do they override it completely?

Comment (BarryThePenguin) [reply .eslintrc.js:58]: I believe they get merged, in the order they are defined. Though the specifics are pretty vague..

- Mentioned briefly in the [eslint migration guide](https://eslint.org/docs/user-guide/migrating-to-7.0.0#lint-files-matched-by-overridesfiles-by-default)
- https://github.com/eslint/rfcs/blob/master/designs/2019-additional-lint-targets/README.md

You can run `npx eslint --print-config ./packages/@ember/string/index.ts` to see what the eslint config for that file will look like..

```json
{
  "env": {},
  "globals": {},
  "parser": "./node_modules/@typescript-eslint/parser/dist/parser.js",
  "parserOptions": {
    "sourceType": "module",
    "project": "./tsconfig.json",
    "tsconfigRootDir": "."
  },
  "plugins": [
    "prettier",
    "disable-features",
    "qunit",
    "import",
    "ember-internal",
    "@typescript-eslint"
  ],
  "rules": {
    ...
  },
  "settings": {
    ...
  },
  "ignorePatterns": [
    ...
  ]
}
```

..compared with the config for a `.js` file `npx eslint --print-config packages/@ember/application/index.js`

```json
{
  "env": {},
  "globals": {
    "window": true,
    "document": true,
    "setTimeout": true,
    "clearTimeout": true,
    "setInterval": true,
    "clearInterval": true,
    "console": true,
    "Map": true,
    "Set": true,
    "Symbol": true,
    "WeakMap": true
  },
  "parser": "./node_modules/babel-eslint/lib/index.js",
  "parserOptions": {
    "ecmaVersion": 2017,
    "sourceType": "module"
  },
  "plugins": [
    "prettier",
    "disable-features",
    "qunit",
    "import",
    "ember-internal"
  ],
  "rules": {
    ...
  },
  "settings": {
    ...
  },
  "ignorePatterns": [
    ...
  ]
}
```

Comment (BarryThePenguin) [reply .eslintrc.js:58]: Looking at this now.. the overrides for `files: [ 'packages/**/*.js' ]` may need to be updated to include `packages/**/*.ts`?

**Review by chancancode — APPROVED** [review]
Generally looks good to me, thanks for working on this! I'll let @rwjblue do a final review since he is probably more familiar with these tools and how we want them set up.

**Comment by chancancode** [comment]
In the meantime, can you rebase on master (to ensure the recent changes didn't cause any new issues)? Thanks again 🙇

File: .prettierrc.js [hunk 9]
```diff
@@ -1,6 +1,7 @@
 'use strict';
 
 module.exports = {
+  arrowParens: 'avoid',
```
Comment (rwjblue) [inline .prettierrc.js:4]: IMHO, we should not override this. The default new Ember app blueprint will not include this customization (based on https://github.com/emberjs/rfcs/blob/master/text/0628-prettier.md), and we shouldn't arbitrarily differ from that.

File: bin/build-for-publishing.js (hunk 8 above)
Comment (rwjblue) [reply bin/build-for-publishing.js]: HA! I just commented the opposite (https://github.com/emberjs/ember.js/pull/19154#discussion_r495614099). Sorry!

@BarryThePenguin - Would mind reverting the `avoid` change, and collapsing the non `--fix` changes to a single commit; then run `yarn eslint --fix .` and commit _just_ that in a second commit? I'll review the first (but trust `eslint --fix` to the right thing for the rest), then we can land...

Comment (BarryThePenguin) [reply bin/build-for-publishing.js:90]: No worries 😅 That should be done now

**Comment by BarryThePenguin** [comment]
@rwjblue Just rebased on master. I've recreated the second commit with `eslint --fix`, instead of resolving conflicts

File: .eslintrc.js (hunk 4 above)
Comment (BarryThePenguin) [reply .eslintrc.js]: I've removed the `@typescript-eslint` changes as #19166 made it difficult to keep them, without a lot of changes. I can reintroduce it in a different PR

**Comment by BarryThePenguin** [comment]
@chancancode @rwjblue is this something you'd still like to see happen? I understand it's difficult to merge as it would impact other open PRs. Don't want it to be a burden for anyone 😅

**Comment by rwjblue** [comment]
> @chancancode @rwjblue is this something you'd still like to see happen? I understand it's difficult to merge as it would impact other open PRs. Don't want it to be a burden for anyone 😅

Yes!!! Sorry for the lack of communication here, I plan to land this this morning (likely by dropping the `--fix` commit and reapplying, then merging locally and pushing to master).

**Comment by rwjblue** [comment]
Merged in 058787de09963bb5fe499e0bb3acd33d3849802b, thank you @BarryThePenguin!

**Pull request closed without merging**